compose/service.py

```diff
@@ -241,29 +245,15 @@ def recreate_container(self, container, **override_options):
             else:
                 raise
 
-        intermediate_container = Container.create(
-            self.client,
-            image=container.image,
-            entrypoint=['/bin/echo'],
-            command=[],
-            detach=True,
-            host_config=create_host_config(volumes_from=[container.id]),
-        )
-        intermediate_container.start()
-        intermediate_container.wait()
-        container.remove()
-
-        options = dict(override_options)
+        # Use a hopefully unique container name by prepending the short id
+        self.client.rename(container.id, container.short_id + container.name)
```

Comment:
I think there is one possible failure case here. If compose crashes between this line , and the `remove()` line, the old container will be left with the temporary name.

The next time compose runs, it wont recreate (since the container has a different name now), and any volumes will be new instances.  The old volumes will still be attached to the temporary container, so they shouldn't be lost.